### PR TITLE
add schema migration

### PR DESCRIPTION
**Pull request by siyao-Siyang**
### **1. Enable schema migration for SQL DB**
**First time to table selection page, schema migration shows 'Not selected'.**
image

**If schema migration is enabled, assessment results will be checked. If any assessment issues that might fail schema migration are found, it will show the assessment results link to check the details.
If schema migration is not enabled, there are two cases. If there are missing tables on the target but users don't select 'migrate schema to target' option, it will show 'Not selected'. If there are no missing tables on target, it will show 'Not needed'.**
image

**There are three tabs after click 'Edit'：**
**The source tables exist on the target and have rows:**
image
**The source tables don't exist on the target and have rows:**
image
**The source tables dont have rows. These tables are not available to select for data migration. But if they dont exist on the target, schema migration is available.**
image

### 2. Auto-mapping source database to target database.
**Find if one of target database has the same name and collation to the selected source database, then this target database is set as dropdown component default value.**
image

### 3. Integrate the assessment results with schema migration
**User can click the 'Assessment results' link to check what issues might fail the schema migration**
image
image

### 4. There are five scenarios:
**4. 1. Full schema on target (Data only migration)**
"Migrate schema to target" checkbox is disable.
image
image
image

**4.2. Full schema missing on target** 
image
image

**4.3. Partially schema missing on target**
image

**4.4. Failed migration**
image

**4.5. All source tables are empty.** 
If they don't exist on the target, the schema migration is still available but not available to select for data migration
image

### 5. Migration details list
**Migration type: Schema only, Data only, Schema and data, Backup and restore**
image

### CODE REVIEW

File: extensions/sql-migration/src/constants/helper.ts [hunk 1]
```diff
@@ -360,3 +360,26 @@ export function selectDatabasesFromList(selectedDbs: string[], databaseTableValu
...
+	var targetType = getMigrationTargetTypeEnum(migration);
+	if (targetType === MigrationTargetType.SQLMI || targetType === MigrationTargetType.SQLVM) {
+		return loc.BACKUP_AND_RESTORE;
+	}
+
+	var enableSchema = migration?.properties?.sqlSchemaMigrationConfiguration?.enableSchemaMigration ?? false;
+	var enableData = migration?.properties?.sqlDataMigrationConfiguration?.enableDataMigration ?? false;
+	return enableSchema && enableData
+		? loc.SCHEMA_AND_DATA
+		: enableSchema ? loc.SCHEMA_ONLY : loc.DATA_ONLY
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/helper.ts:375]: missing a " ; " at the end

File: extensions/sql-migration/src/constants/strings.ts [hunk 2]
```diff
@@ -1709,3 +1717,82 @@ export function TDE_MIGRATE_STATUS_ERROR(state: string, error: string): string {
...
+export const MIGRATION_TYPE = localize('sql.migration.schema.type', "Migration type");
+export const SCRIPT_DEPLOYMENT = localize('sql.migration.schema.script.deployment', "Script deployment");
+export const DEPLOYMENT_STARTED = localize('sql.migration.schema.script.deployment.started', "Deployment started");
+export const DEPLOYMENT_ENDED = localize('sql.migration.schema.script.deployment.ended', "Deployment ended");
+export const DEPLOYMENT_COUNT = localize('sql.migration.schema.script.deployment.count', "Deployment count");
+export const DEPLOYMENT_ERROR_COUNT = localize('sql.migration.schema.script.deployment.error.count', "Deployment error count");
+export const SCHEMA_MIGRATION_ASSESSMENT_WARNING_MESSAGE = localize('sql.migration.schema.assessment.warning.message', "The detected issues shown below might fail the schema migration. Some of them might be entirely unsupported and the others might be partially supported in Azure SQL Database. \nTherefore, please review the assessment results and make sure all of the issues will not fail the schema migration.\nHowever, it is allowed to proceed the schema migration and DMS will migrate the objects as possible as it can.");
+
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:1751]: Please avoid using product acronyms like "DMS" as it dilutes the value/name of the service... instead it is recommended that we refer to the service by name "Database Migration Service" or "Azure Database Migration Service"

File: extensions/sql-migration/src/dialog/tableMigrationSelection/schemaMigrationAssessmentDialog.ts [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+	}
+
+	private async _initializeDialog(dialog: azdata.window.Dialog): Promise<void> {
+		dialog.registerContent(async (view: azdata.ModelView) => {
+			dialog.message = {
+				level: azdata.window.MessageLevel.Warning,
+				text: constants.SCHEMA_MIGRATION_ASSESSMENT_WARNING_MESSAGE,
+			};
+			dialog.okButton.hidden = true;
+			dialog.cancelButton.label = "Close";
```
Comment (brian-harris) [inline extensions/sql-migration/src/dialog/tableMigrationSelection/schemaMigrationAssessmentDialog.ts:70]: "Close" should be added as a resource string

File: extensions/sql-migration/src/dashboard/migrationsListTab.ts [hunk 4]
```diff
@@ -346,6 +348,13 @@ export class MigrationsListTab extends TabBase<MigrationsListTab> {
 						getMigrationMode(m2),
 						sortDir));
 				return;
+			case TableColumns.type:
+				migrations.sort(
+					(m1, m2) => this.stringCompare(
+						getMigrationType(m1),
+						getMigrationType(m2),
+						sortDir));
+				return;
```
Comment (brian-harris) [inline extensions/sql-migration/src/dashboard/migrationsListTab.ts:357]: I wonder if adding "Type" column here is needed as it does not apply across all migration types.  Schema/data migration only applies to SQL DB and empty for SQL MI/SQL VM.

**Review by stuti149 — APPROVED** [review]

File: extensions/sql-migration/src/dashboard/migrationsListTab.ts (hunk 4 above)
Comment (siyao-Siyang) [reply extensions/sql-migration/src/dashboard/migrationsListTab.ts:357]: ok. maybe remove it. For SQLMI/VM, it shows "Backup - restore"

Comment (siyao-Siyang) [reply extensions/sql-migration/src/dashboard/migrationsListTab.ts:357]: Confirmed with Abhishek. it is ok to include this column.

**Pull request merged**